File: backend/app/api/attention_proxy.py
"""Proxy attention-service routes through the main API (single public URL for Vercel + tunnels)."""
import os

import httpx
from fastapi import APIRouter, UploadFile, File, Form, Header, HTTPException

router = APIRouter()

ATTENTION_BASE = os.getenv("ATTENTION_SERVICE_URL", "http://127.0.0.1:8001").rstrip("/")
VOICE_TIMEOUT = float(os.getenv("ATTENTION_VOICE_TIMEOUT", "120"))


async def _forward_multipart(
    path: str,
    form_data: dict,
    files: dict,
    authorization: str | None,
) -> dict:
    headers = {}
    if authorization:
        headers["Authorization"] = authorization

    try:
        async with httpx.AsyncClient(timeout=VOICE_TIMEOUT) as client:
            response = await client.post(
                f"{ATTENTION_BASE}{path}",
                data=form_data,
                files=files,
                headers=headers,
            )
    except httpx.RequestError as exc:
        raise HTTPException(
            status_code=503,
            detail=f"Attention service unavailable: {exc}",
        ) from exc

    if response.status_code >= 400:
        detail = response.text
        try:
            detail = response.json().get("detail", detail)
        except Exception:
            pass
        raise HTTPException(status_code=response.status_code, detail=detail)

    return response.json()


@router.post("/voice/register")
async def proxy_register_voice(
    candidate_id: str = Form(...),
    audio: UploadFile = File(...),
    authorization: str | None = Header(None),
):
    audio_bytes = await audio.read()
    return await _forward_multipart(
        "/voice/register",
        {"candidate_id": candidate_id},
        {
            "audio": (
                audio.filename or "voice.webm",
                audio_bytes,
                audio.content_type or "audio/webm",
            )
        },
        authorization,
    )


@router.post("/voice/verify")
async def proxy_verify_voice(
    candidate_id: str = Form(...),
    audio: UploadFile = File(...),
    authorization: str | None = Header(None),
):
    audio_bytes = await audio.read()
    return await _forward_multipart(
        "/voice/verify",
        {"candidate_id": candidate_id},
        {
            "audio": (
                audio.filename or "voice.webm",
                audio_bytes,
                audio.content_type or "audio/webm",
            )
        },
        authorization,
    )


@router.post("/attention/analyze")
async def proxy_analyze_attention(
    file: UploadFile = File(...),
):
    file_bytes = await file.read()
    try:
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                f"{ATTENTION_BASE}/attention/analyze",
                files={
                    "file": (
                        file.filename or "frame.jpg",
                        file_bytes,
                        file.content_type or "image/jpeg",
                    )
                },
            )
    except httpx.RequestError as exc:
        raise HTTPException(
            status_code=503,
            detail=f"Attention service unavailable: {exc}",
        ) from exc

    if response.status_code >= 400:
        raise HTTPException(status_code=response.status_code, detail=response.text)

    return response.json()

File: backend-node/src/routes/attentionProxy.js
import { Router } from 'express'
import multer from 'multer'
import { env } from '../config/env.js'
import { asyncHandler } from '../middleware/errorHandler.js'
import { httpError } from '../utils/errors.js'

const router = Router()
const upload = multer({ storage: multer.memoryStorage() })

async function forwardMultipart(path, fields, fileField, authorization) {
  const form = new FormData()
  for (const [key, value] of Object.entries(fields)) {
    form.append(key, value)
  }
  if (fileField) {
    form.append(
      fileField.name,
      new Blob([fileField.buffer], { type: fileField.contentType }),
      fileField.filename,
    )
  }

  const headers = {}
  if (authorization) headers.Authorization = authorization

  let response
  try {
    response = await fetch(`${env.attentionServiceUrl}${path}`, {
      method: 'POST',
      headers,
      body: form,
      signal: AbortSignal.timeout(env.attentionVoiceTimeoutMs),
    })
  } catch (err) {
    throw httpError(503, `Attention service unavailable: ${err.message}`)
  }

  const text = await response.text()
  let data
  try {
    data = JSON.parse(text)
  } catch {
    data = text
  }

  if (response.status >= 400) {
    const detail =
      data && typeof data === 'object' && 'detail' in data ? data.detail : text
    throw httpError(response.status, detail)
  }

  return data
}

router.post(
  '/voice/register',
  upload.single('audio'),
  asyncHandler(async (req, res) => {
    const candidateId = req.body?.candidate_id
    if (!candidateId) throw httpError(422, [{ msg: 'candidate_id is required' }])
    if (!req.file) throw httpError(422, [{ msg: 'audio is required' }])

    const data = await forwardMultipart(
      '/voice/register',
      { candidate_id: candidateId },
      {
        name: 'audio',
        buffer: req.file.buffer,
        filename: req.file.originalname || 'voice.webm',
        contentType: req.file.mimetype || 'audio/webm',
      },
      req.headers.authorization,
    )
    res.json(data)
  }),
)

router.post(
  '/voice/verify',
  upload.single('audio'),
  asyncHandler(async (req, res) => {
    const candidateId = req.body?.candidate_id
    if (!candidateId) throw httpError(422, [{ msg: 'candidate_id is required' }])
    if (!req.file) throw httpError(422, [{ msg: 'audio is required' }])

    const data = await forwardMultipart(
      '/voice/verify',
      { candidate_id: candidateId },
      {
        name: 'audio',
        buffer: req.file.buffer,
        filename: req.file.originalname || 'voice.webm',
        contentType: req.file.mimetype || 'audio/webm',
      },
      req.headers.authorization,
    )
    res.json(data)
  }),
)

router.post(
  '/attention/analyze',
  upload.single('file'),
  asyncHandler(async (req, res) => {
    if (!req.file) throw httpError(422, [{ msg: 'file is required' }])

    const form = new FormData()
    form.append(
      'file',
      new Blob([req.file.buffer], {
        type: req.file.mimetype || 'image/jpeg',
      }),
      req.file.originalname || 'frame.jpg',
    )

    let response
    try {
      response = await fetch(`${env.attentionServiceUrl}/attention/analyze`, {
        method: 'POST',
        body: form,
        signal: AbortSignal.timeout(60000),
      })
    } catch (err) {
      throw httpError(503, `Attention service unavailable: ${err.message}`)
    }

    const text = await response.text()
    if (response.status >= 400) {
      throw httpError(response.status, text)
    }
    res.json(JSON.parse(text))
  }),
)

export default router
